---
title: "Holger Russ scRNA-Seq analysis expt2."
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    fig_caption: yes
    fig_retina: 1 
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source("../../R/globals.R")
data_dir_expt1 <- "~/Projects/10x_data/runs/20180524_russ/results"
data_dir_expt2 <- "~/Projects/10x_data/runs/20180918_russ/results"
tcols <- rev(brewer.pal(11, "RdGy")[c(1:5, 7)])[1:6]
```

## Experiment Summary

Sample | Description
------------- | ---------------- 
Holger_CTRL | hPSC derived Beta cells 
Holger_T1D_modeling | hPSC derived Beta Cells exposed to diabetes related stress
Russ_GFP_Sort | 2nd experiment
Russ_d36GFP_Reagg | 2nd experiment

```{r get_data}

samples_expt1 <- c(
  "Holger_CTRL",
  "Holger_T1D_modeling")

sample_paths_1 <- file.path(data_dir_expt1,
                          samples_expt1,
                          "outs", 
                          "filtered_gene_bc_matrices", 
                          "GRCh38_transgenes")

names(sample_paths_1) <- samples_expt1

samples_expt2 <- c(
  "Russ_GFP_Sort",
  "Russ_d36GFP_Reagg")

sample_paths_2 <- file.path(data_dir_expt2,
                          samples_expt2,
                          "outs", 
                          "filtered_gene_bc_matrices", 
                          "GRCh38_transgenes")

names(sample_paths_2) <- samples_expt2

sample_paths <- c(sample_paths_1, sample_paths_2)
bcells <- Seurat::Read10X(sample_paths)
```

```{r create_seurat, message = F, results = 'hide', warning = F}
bcells <- CreateSeuratObject(raw.data = bcells, 
                              min.cells = 5,
                              min.genes = 250, 
                              project = "Russ",
                              names.field = c(1,2), 
                              names.delim = "_")
```

```{r additional_mdata}

# add correct sample identifier
ids <- bcells@cell.names
ids <- str_split(ids, "_[ATCG]{10}", simplify = T) %>% .[, 1]
ids <- data.frame(row.names = bcells@cell.names,
                  expt = ids)
bcells <- AddMetaData(bcells, ids)
```

## QC Metrics
  
```{r examine_umi_and_ngenes, fig.width = 10, fig.height = 8, warning = F}
summary_dat <- data_frame(cell = colnames(bcells@raw.data), 
                          expt = str_split(colnames(bcells@raw.data), "_[ATCG]{10}", simplify = T) %>% .[, 1],
                          nUMIs = colSums(as.matrix(bcells@raw.data)),
                          nGenes = colSums(as.matrix(bcells@raw.data) > 0))

summary_dat <- summary_dat %>% 
  group_by(expt) %>% 
  dplyr::arrange(desc(nUMIs))

n_cells <- summary_dat %>%  
  group_by(expt) %>%  
  summarize(n_cells = n(),
            median_UMIs = median(nUMIs),
            median_Genes = median(nGenes))

knitr::kable(n_cells,
             caption = "Number of cells prior to filtering and summary statistics")

a <- ggplot(summary_dat, aes(expt, nUMIs)) +
  geom_violin(aes(fill = expt)) +
  scale_fill_brewer(palette = "Paired") +
  ylab("Number of UMIs (Log10)") +
  theme(axis.text.x = element_blank())

b <- ggplot(summary_dat, aes(expt, nGenes)) +
    geom_violin(aes(fill = expt)) +
  scale_fill_brewer(palette = "Paired") +
  ylab("Number of Genes (at least 1 UMI) Log10") +
  theme(axis.text.x = element_blank())
plot_grid(a, b, ncol = 1)
```

Overall the numbers of UMIs and genes detected are good and in the expected range for single cell data. 

```{r write_matrix, eval = F}
dir.create("count_matrix", showWarnings = FALSE)

write.table(as.matrix(bcells@raw.data), 
            file.path("count_matrix", "all_samples_raw_counts_matrix.txt"),
            quote = F, sep = "\t")

gzip(file.path("count_matrix", 
               "all_samples_raw_counts_matrix.txt"),
     overwrite = TRUE,
     remove = TRUE)
```

```{r utility_fxn}
cell_counts <- function(seurat_object){
  #returns tbl_df with number of cells per sample type in seurat object
  map_df(levels(seurat_object@ident), 
         ~data_frame(sample = .x, 
                     n_cells = length(WhichCells(seurat_object, ident = .x))))
}
```

## Mitochondrial Reads {.tabset .tabset-fade}

The percentage of reads aligning to mitochondria is often an indicator of cell quality. Cells that have lysed prior to capture preferentially retain mitochondial RNA, but lose cellular cytoplasmic mRNA. As shown below there is a strong negative correlation with the number of UMIs and the proportion of reads aligning to mitochondria. Cells with less than 20% mitochondrial reads will be retained for further analysis.

```{r QC}
mito.genes <- grep("^MT-", rownames(bcells@raw.data), value = T)
proportion.mito <- Matrix::colSums(bcells@raw.data[mito.genes, ]) /
  Matrix::colSums(bcells@raw.data)

bcells <- AddMetaData(bcells, proportion.mito, "proportion.mito")

unfiltered <- bcells@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(summary_dat, by = c("cell", "expt")) %>% 
  as_data_frame()

ggplot(unfiltered, aes(expt,
                       proportion.mito)) +
  geom_jitter(size = 0.25) +
  geom_violin(aes(fill = expt), alpha = 0.66) +
  scale_fill_brewer(palette = "Paired") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1),
    axis.title.x = element_blank()
  )
```

### All Samples

```{r sample_ids}
ggplot(unfiltered, 
       aes(nUMI, proportion.mito)) +
  geom_point(aes(color = expt)) +
  scale_color_brewer(palette = "Paired")
```

```{r, results ='asis'}

sample_names <- as.character(unique(bcells@meta.data$expt))
per_sample <- map(sample_names, ~filter(unfiltered, 
                                        expt == .x))

# generate tab with individual plot programmatically 
# see https://stackoverflow.com/questions/43752095/programmatically-insert-header-and-plot-in-same-code-chunk-with-r-markdown-using?noredirect=1&lq=1

for(i in seq_along(per_sample)){
  .col <- brewer.pal(10, "Paired")[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nUMIs, proportion.mito)) +
        geom_point(aes(color = expt)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```

Cells with low UMIs tend to have a higher percentage of mitochondrial reads, suggesting that these cells have lysed prematurely. However, overall a large proportion of cells are retained for further analysis after removing cells with > 20% UMIs derived from mitochondiral transcripts. 

## Filtered Data {.tabset .tabset-fade}

After filtering the distribution of UMIs, Genes, and Proportion of mitochondrial reads are shown as selectable tabs. 

```{r, apply_filters}
bcells <- FilterCells(bcells, 
                       subset.names = c("nUMI", "proportion.mito"), 
                       low.thresholds = c(250, -Inf), 
                       high.thresholds = c(75000, 0.20))
```

```{r ngenes_filter}
knitr::kable(cell_counts(SetAllIdent(bcells, "expt")),
             caption = "Number of cells passing an additional filter requiring less than 20% mitochondrial reads and at least 250 genes detected, and UMI < 75000")
```

### Number of Genes
```{r ngenes_post_filter}
VlnPlot(bcells, "nGene", 
        cols.use = brewer.pal(10, "Paired"),
        group.by = "expt",
        x.lab.rot = T)
```

### Number of UMIs
```{r}
VlnPlot(bcells, "nUMI", 
        cols.use = brewer.pal(10, "Paired"),
        group.by = "expt",
        x.lab.rot = T)
```

### Proportion of Mitochondria
```{r}
VlnPlot(bcells, "proportion.mito", 
        cols.use = brewer.pal(10, "Paired"),
        group.by = "expt",
        x.lab.rot = T)

```

### GFP/LUC counts
```{r GFP_counts}

VlnPlot(bcells, 
        c("egfp", "luc-firefly"),
        cols.use = brewer.pal(n = 10, "Paired"), 
        group.by = "expt",
        x.lab.rot = T,
        y.log = T)

```

## Preprocess the data for PCA, TSNE, and Clustering 
  The raw UMI count matrix will next be preprocessed for downstream clustering and visualization
  
### Normalize the data
  First the UMI count data will be normlized. The normalization strategy implements a pooled normalization approach to improve standard normalization [Lun2016-fe]. Different cell populations can have large gene abundance and RNA content differences, which is problematic for simple library-size based normalization (i.e # of UMIs per gene / total UMIs per cell). This pooling approach find clusters cells to identify the cell popuations, then computes per cell normalization factors. The R package `scran`[http://bioconductor.org/packages/release/bioc/html/scran.html] implements this normalization method. 

```{r load_seurat}
bcells <- readRDS("sobj_nocca.rds")
bcells <- SetAllIdent(bcells, "orig.ident")
```

```{r normalize, results = 'hide', eval = F}
bcells <- NormalizeData(bcells)

```

### Identify variable genes

  The normalized UMI count matrix contains `r  length(rownames(bcells@data))` genes and `r length(colnames(bcells@data))` cells. Most of these genes have very low expression or very low variance across all of the cells in the dataset. These genes are therefore very uninformative for classifying different cell types. In this next step the set of variable genes useful for classification will be identified, and used for downstream PCA, TSNE, and classification purposes. 
  
  Variable genes will be selected by examining a plot of the average expression across all cells on the x axis to the dispersion across all genes. Highly variable genes will have non-zero dispersion. The dispersion is defined by the variance / mean. 
  
```{r get_variable_genes, results = 'hide', eval = F}
bcells <- FindVariableGenes(bcells, do.plot = F)

length(bcells@var.genes)

vargenes <- rownames(bcells@hvg.info)[1:2000]
```

### Control for noise
  Single-cell data is often noisy, and it is difficult to compare cells with very high UMIs to very low UMIs. For classification we do not want to bias our clusters based on nuisance variables such as UMI count or proportion of mitochondria. To help mitigate bias in our clustering a regression model is fit to predict gene expression based on nuisance variables. The residuals from this fit are then scaled and centered (i.e. turned into z-scores) for classification purposes. The scaled regressed data is not used for differential expression, but used only for classifcation, TSNE visualization, and pseudotime analysis. 

```{r scale_data, results='hide', message = F, eval = F}
bcells <- ScaleData(bcells, genes.use = vargenes, vars.to.regress = c("nUMI", "proportion.mito"))
```

### Calculate Principal Components

  PCA will now be performed on processed data using only the set of variable genes defined above. PCA will generate a set of new vectors that describe the variablity in the data. A subset of these principal components will be selected to be used as input data for TSNE visualization. Selecting the proper number of components is informed by examining the amount of variance captured in each PC. Shown below is a plot of variance captured in each PC. At a certain point, additional PCs capture little additional varaince, and therefore are uninformative for classification. For this dataset 17 PCs will be used for classification.
  
```{r PCA_calc, results='hide', eval = F}
bcells <- RunPCA(bcells, 
               pc.genes = vargenes, 
               do.print = FALSE, 
               pcs.print = 1:5, 
               genes.print = 5,
               seed.use = 20180530)

PCElbowPlot(bcells)
```

### PCA Plots {.tabset .tabsetfade}
  Shown in the tabs are 2 dimensional plots showing different principal components. Overall the data looks good as there is quite alot of structure to the data and many consisent populations that match the sample types are observed.
  
```{r pc_plots, results ='asis'}
pcs <- list(
  c(1, 2),
  c(1, 3),
  c(1, 4),
  c(2, 3),
  c(2, 4),
  c(3, 4)
)


for(i in seq_along(pcs)){
  cat('\n#### ', 'PC', pcs[[i]][1], ' vs PC', pcs[[i]][2], '\n', sep = "")
  PCAPlot(bcells, 
               pcs[[i]][1], 
               pcs[[i]][2], 
               cols.use = brewer.pal(10, "Paired"), 
        group.by = "expt")
  #print(p)
  cat('\n')
}
```

### Run TSNE {.tabset .tabsetfade}
  [TSNE](https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding) is a non-linear method that is used to visualize the relationships between cells in two dimensions. TSNE will use the first 17 PCs calculated above to project cells into 2D space for visualization. The TSNE plot will be used for visualization purposes, but note that clustering will be performed using the 17 PCs rather than just the 2 dimensions of the TSNE. The perplexity setting of the TSNE analysis can impact the spacing of points and for this data a default setting of 30 (perplexity = 30) works well. Interpreting a tSNE plot can be somewhat challenging, a group from google produced a nice [report](https://distill.pub/2016/misread-tsne/) that discusses some of the potential issues with tSNE.
  
    
```{r select_dims, eval = F}
bcells <- RunTSNE(bcells, 
                   dims.use = 1:17, 
                   do.fast = T, 
                   k.seed = 20180530,
                   perplexity = 30)
```
  
  
#### Sample Names

```{r plot_TSNE, fig.width = 10, fig.height = 10}
plt <- TSNEPlot(bcells, 
         colors.use = brewer.pal(n = 10, "Paired"), 
         group.by = "expt", pt.size = 0.2)

save_plot("tsne_no_cca.pdf", plt, base_height = 6)
```


#### Genes Detected
```{r, fig.width = 10, fig.height = 10 }
FeaturePlot(bcells, "nGene", no.legend = F)
```

#### UMIs Detected
```{r, fig.width = 10, fig.height = 10 }
FeaturePlot(bcells, "nUMI", no.legend = F)
```

#### Proportion Mito
```{r, fig.width = 10, fig.height = 10 }
FeaturePlot(bcells, "proportion.mito", no.legend = F)

```

#### eGFP
```{r, fig.width = 10, fig.height = 10 }
FeaturePlot(bcells, "egfp", no.legend = F)
```


#### firefly-luc
```{r, fig.width = 10, fig.height = 10 }
FeaturePlot(bcells, "luc-firefly", no.legend = F)
```

#### Per expt
```{r, fig.width = 16, fig.height= 4}
bcells <- SetAllIdent(bcells, "expt")
sample_names <- unique(bcells@meta.data$expt)
ord <- levels(bcells@meta.data$expt)

cell_ids <- map(ord, ~bcells@cell.names[bcells@meta.data$expt == .x])
names(cell_ids) <- ord
.cols <- brewer.pal(n = 10, "Paired")[1:length(sample_names)]

p <- map2(cell_ids,
         .cols,
         ~TSNEPlot(bcells, 
                   do.label = F, 
                   pt.size = 0.5, 
                   cells.use = .x, 
                   do.return = T, 
                   colors.use = .y,
                   label.size = 5))

p[[5]] <- TSNEPlot(bcells, 
                   do.label = F, 
                   pt.size = 0.5, 
                   do.return = T, 
                   colors.use  = .cols,
                   label.size = 5)


plt <- plot_grid(plotlist = p, nrow = 1, 
                 align = "hv")

plt

save_plot("tsne_per_expt_no_cca.pdf", plt, nrow = 1, ncol = 5)
```

## Marker gene expression for each cluster {.tabset .tabset-fade}

  Plotted below are a set of known marker genes that help to identify each cell population. 

Click on the tabs to view other markers

```{r marker genes, fig.width = 10, fig.height= 10}

control_markers <- c(
  "CD9",
  "ST8SIA1",
  "INS",
  "NPY",
  "KCNE4",
  "HCN1",
  "DAB1",
  "ST8SIA1",
  "HCN4",
  "NEUROD1",
  "SCGB2A1",
  "KLHL1",
  "SLC44A4",
  "egfp",
  "FEV"
)
```

### Various
```{r, fig.width=10, fig.height=20}
FeaturePlot(bcells, control_markers, 
            pt.size = 0.50, cols.use = tcols)

```


## Gene expression per expt

Next we will determine genes enriched in each sample by comparing gene expression in each sample to all other samples. Plotted below are the top 10 enriched genes detected in each sample for all cell transcriptomes (Yellow is high, purple is low).

```{r find_markers}
bcells <- SetAllIdent(bcells, "expt")
if (!file.exists("per_sample_markers.txt")){
  bcells_markers <- FindAllMarkers(bcells, 
                                  only.pos = TRUE,  
                                  min.pct = 0.25, 
                                  thresh.use = 0.25)
  
  write_tsv(bcells_markers, "per_sample_markers.txt")
} else {
  bcells_markers <- read_tsv("per_sample_markers.txt")
}

bcells_markers %>% 
  group_by(cluster) %>% 
  top_n(25, avg_logFC) -> top25

```


### xcel spreadsheet

```{r}
markers <- read_tsv("per_sample_markers.txt")

list_of_markers <- split(markers, markers$cluster)
names(list_of_markers) <- str_c("cluster_", names(list_of_markers))
list_of_markers <- map(list_of_markers, 
                       ~set_xlsx_class(.x, "gene", "Text"))

readme_sheet <- data_frame(
  Columns = c(
  "Genes differentially expressed between each sample and all other cells",
  "",
  "Columns",
  "pval",
  "avg_logFC",
  "pct.1",
  "pct.2",
  "p_val_adj",
  "cluster",
  "gene"
), Description = c(
  "",
  "",
  "",
  "p-value from wilcox test of indicated cluster compared to other clusters",
  "average fold change expressed in natural log",
  "percent of cells expressing gene (UMI > 0) in cluster",
  "percent of cell expressing gene (UMI > 0) in all other clusters",
  "Bonferroni corrected p-value",
  "cluster name",
  "gene name"
))

readme_sheet <- list(README = readme_sheet)
names(readme_sheet) <- "README"

openxlsx::write.xlsx(c(readme_sheet, 
                       list_of_markers),
                     "markers_per_sample.xlsx")
```

```