---
title: "Compare to distinct beta-cell subtypes"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
bibliography: "/Users/kriemo/Projects/russ/docs/bibliography.bib"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    fig_caption: yes
    fig_retina: 1 
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source("../../R/globals.R")
tcols <- rev(brewer.pal(11, "RdGy")[c(1:5, 7)])[c(1,6)]

bcells <- readRDS(file.path(results_dir, 
                            "2018-05-30", 
                            "sobj.rds"))
```

# Compare to beta-cell subtype expression profiles

Previous work by Holger and collegues identified four distinct subpopulations of human beta-cells ([paper](https://doi.org/10.1038/ncomms11756)). RNA-Seq was performed to identify the gene expression signatures of these subtypes. To gain insight into the scRNA-seq data I will compare the bulk RNA-Seq data to the single cell RNA-Seq data. The [GEO dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE80780) contains a counts table, which will be used for the initial analysis. I have also downloaded the raw data and begun to process the data into a counts table with hg38 annotations (the GEO counts table is hg19).

```{r dl_counts_table}
dl_fn <- file.path(docs_dir, "extdata", "GSE80780_counts.tsv.gz")
dl_url <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE80780&format=file&file=GSE80780%5Freads%2Etsv%2Egz"

if(!file.exists(dl_fn)){
  download.file(dl_url, dl_fn)
}

count_dat <- read_tsv(dl_fn)

# the best.rep column seems to indicate the best transcript to represent the gene, as if you filter for best.rep == 1 you get back the # of unique gene names

count_dat <- filter(count_dat, best.rep == 1)

count_dat
```

```{r normalize_bulk_data}
# average lib-sized normalized counts for bulk RNA-Seq
norm_counts <- count_dat %>% 
  select(-best.rep, -Transcript) %>% 
  gather(sample, count, -Gene) %>% 
  separate(sample, c("donor", "subtype"), sep = "_") %>% 
  group_by(donor, subtype) %>% 
  mutate(total_counts = sum(count),
         norm_counts = 1e6 * (count / total_counts)) %>% 
  group_by(Gene, subtype) %>%
  summarize(avg_norm_counts = mean(norm_counts)) %>% 
  spread(subtype, avg_norm_counts) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("Gene") 

# run deseq with LRT test to find genes de between subtypes
count_mat <- count_dat %>% 
  select(-Transcript, -best.rep) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("Gene")

pdata <- data.frame(donor = str_split(colnames(count_mat), 
                                      "_", 
                                      simplify = T)[, 1],
                    subtype = str_split(colnames(count_mat), 
                                      "_", 
                                      simplify = T)[, 2],
                    row.names = colnames(count_mat))

library(DESeq2)
dds <- DESeqDataSetFromMatrix(count_mat, 
                              colData = pdata,
                              design = ~ subtype + donor)

dds <- DESeq(dds, test = "LRT", reduced = ~ donor)
dds_res <- results(dds, tidy = T)
dds_sig <- dds_res %>% 
  filter(padj < 0.1) %>% 
  pull(row)
```

```{r compare to scRNA}

# select variable genes
markers <- suppressMessages(read_tsv("../2018-05-30/res0.3_markers.txt"))

bcells_markers <- markers %>% pull(gene)

bcells <- FindVariableGenes(bcells, 
                    mean.function = ExpMean, 
                    dispersion.function = LogVMR, 
                    x.low.cutoff = 0.1, 
                    x.high.cutoff = 5, 
                    y.cutoff = 0.25, do.contour = F, do.plot = T)

var_genes <- intersect(intersect(rownames(norm_counts),
                       dds_sig),
                       rownames(bcells@data))

cor_comp <- cor(
    as.matrix(bcells@data[var_genes, ]),
    norm_counts[var_genes, ],
    method = "spearman")

cor_comp_dat <- as.data.frame(cor_comp) %>% 
  tibble::rownames_to_column("cell")

plts <- map(colnames(cor_comp_dat)[2:ncol(cor_comp_dat)],
    ~plot_feature(bcells,
                  plot_dat = cor_comp_dat, 
                  plot_col = .x,
                  pt.alpha = 1,
                  pt.size = 0.1))

plt <- plot_grid(plotlist = plts, nrow = 1)
save_plot("cor_based_on_markers.pdf", plt, nrow = 1, ncol = 4)
```

```{r}
# table 1from paper 
tbl_dat <- tibble::tribble(
  ~gene, ~gene_desc, ~log2fc, ~pval, ~fdr,
         "ST8SIA1",             "ST8.alpha-N-acetyl-neuraminide.alpha-2,8-sialyltransferase.1",  "12.52", "7.1E-11",   "0",
           "HCN1", "Hyperpolarization-activated cyclic nucleotide-gated potassium channel 1",  "3.03",  0.00015, 0.03,
           "DAB1",                 "Dab, reelin signal transducer, homologue 1 (Drosophila)",  "2.21",  2.9e-05, 0.01,
          "KCND3",     "Potassium channel, voltage-gated Shal-related subfamily D, member 3",  "2.16",  7.8e-09,    0,
           "SIX3",                                                          "SIX homeobox 3",  "2.01",  1.9e-09,    0,
         "ATP8B1",       "ATPase, aminophospholipid transporter, class I, type 8B, member 1",  "1.98",  0.00042, 0.05,
          "KCNJ8",            "Potassium channel, inwardly rectifying subfamily J, member 8",   "1.9",  0.00068, 0.07,
           "SPP1",                                               "Secreted phosphoprotein 1",  "1.87",  4.7e-09,    0,
           "TFF3",                                           "Trefoil factor 3 (intestinal)",  "1.81",  8.3e-13,    0,
          "ITGA1",                                                       "Integrin, alpha 1",  "1.76",  0.00023, 0.04,
          "KCNE4",  "Potassium channel, voltage-gated subfamily E regulatory beta subunit 4",  "1.76",  0.00064, 0.06,
          "RBM43",                                            "RNA-binding motif protein 43",  "1.57",  0.00035, 0.04,
          "SCN3A",                  "Sodium channel, voltage-gated, type III, alpha subunit",  "1.57",  0.00073, 0.07,
    "ABCC9",                 "ATP-binding cassette, sub-family C (CFTR/MRP), member 9",  "1.54",  3.2e-05, 0.01,
          "SULF2",                                                             "Sulfatase 2",  "1.49",  2.2e-05, 0.01,
           "PON3",                                                           "Paraoxonase 3",  "1.43",  0.00053, 0.06,
            "GSN",                                                                "Gelsolin",   "1.4",  7.3e-06,    0,
          "OXCT1",                                             "3-oxoacid CoA transferase 1", "−1.33",  2.8e-05, 0.01,
          "TMCC3",                           "Transmembrane and coiled-coil domain family 3", "−1.33",  8.2e-05, 0.02,
       "NEUROD1",                                              "Neuronal differentiation 1", "−1.34",    0.001, 0.09,
        "FAM159B",                           "Family with sequence similarity 159, member B", "−1.36",  6.7e-05, 0.02,
          "MAFB",        "v-maf avian musculoaponeurotic fibrosarcoma oncogene homologue B", "−1.38",  0.00014, 0.03,
         "SPTLC2",                  "Serine palmitoyltransferase, long-chain base subunit 2", "−1.38",  0.00022, 0.03,
       "TNFRSF21",                  "Tumour necrosis factor receptor superfamily, member 21", "−1.39",  4.5e-07,    0,
          "G6PC2",                                                 "Glucose-6-phosphatase 2", "−1.46",  3.2e-09,    0,
          "RFX6",                                                  "regulatory factor X, 6", "−1.46",    9e-07,    0,
        "SCGB2A1",                                      "Secretoglobin, family 2A, member 1", "−1.49",  8.9e-09,    0,
  "SLC2A2",     "Solute carrier family 2 (facilitated glucose transporter), member 2", "−1.53",  7.2e-05, 0.02,
           "HCN4", "Hyperpolarization-activated cyclic nucleotide-gated potassium channel 4", "−1.77",  0.00094, 0.08
  )


plts <- map(c("CD9", tbl_dat$gene),
    ~plot_feature(bcells,
                  gene = .x,
                  pt.alpha = 1,
                  pt.size = 0.1))

plt <- plot_grid(plotlist = plts)
save_plot("table_1_markers_tsne.pdf", plt,
          nrow = 5,
          ncol = 6)
```

Next lets take the DE genes identified in the study (CD9+ vs - and ST8 + vs -) and use clustifyr in gene set mode.

```{r}

dl_fn <- c(file.path(docs_dir, "extdata", "cd9_de.tsv.gz"),
           file.path(docs_dir, "extdata", "st8_de.tsv.gz"))
dl_urls <- c(
  "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE80780&format=file&file=GSE80780%5FCD9pm%2Etsv%2Egz",
  "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE80780&format=file&file=GSE80780%5FST8pm%2Etsv%2Egz")

if(!file.exists(dl_fn[1])){
  download.file(dl_urls[1], dl_fn[1])
}
if(!file.exists(dl_fn[2])){
  download.file(dl_urls[2], dl_fn[2])
}

cd9 <- read_tsv(dl_fn[1])
st8 <- read_tsv(dl_fn[2])

sig_genes <- map_dfr(lst(cd9, st8),
                 ~filter(.x,
                         FDR < 0.1), .id = "comp")

sig_genes <- mutate(sig_genes, 
                    class = ifelse(comp == "cd9" & FC > 0,
                                   "cd9+",
                                   ifelse(comp == "cd9" & FC < 0,
                                          "cd9-",
                                          ifelse(comp == "st8" & FC > 0,
                                                 "st8+",
                                                 "st8-"))))

sig_genes <- filter(sig_genes, Gene %in% rownames(bcells@data))
sig_gene_list <- split(sig_genes, sig_genes$class)
sig_gene_list <- imap(sig_gene_list, 
                     function(x, y){
                      df <- data_frame(x$Gene)
                      colnames(df) <- y
                      df})
library(clustifyR)


b <- ifelse(as.matrix(bcells@data[sig_genes$Gene, ]) > 0,
            0,
            1)

res <- map(sig_gene_list, ~compare_lists(b, .x, n = 10000)) 
res <- do.call(cbind, res)

bcell_meta <- GetCellEmbeddings(bcells, "tsne") %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell")

plts <- plot_cor(res,
         bcell_meta,
         colnames(res), cluster = "cell")

plt <- plot_grid(plotlist = plts, nrow = 1)
save_plot("cor_based_on_gene_lists.pdf", plt,
          nrow = 1,
          ncol = 4)
```


## label clusters


```{r}
control_markers <- c(
  "CD9",
  "ST8SIA1",
  "INS",
  "NPY",
  "KCNE4",
  "HCN1",
  "DAB1",
  "ST8SIA1",
  "HCN4",
  "NEUROD1",
  "SCGB2A1",
  "KLHL1",
  "SLC44A4",
  "GCG", # alpha cells glucagon
  "IAPP", # beta cells amylin,
  "SST", #delta cells Somatostatin
  "PPY", #PP cells  (gamma cells or F cells) pancreatic polypeptide
  "GHRL" #Epsilon cells
)

plts <- map(control_markers,
    ~plot_feature(bcells,
                  gene = .x,
                  pt.alpha = 1))

plt <- plot_grid(plotlist = plts)
save_plot("markers.pdf", plt,
          nrow = 4,
          ncol = 5)
```


## get markers from scRNA-seq datasets


```{r}
url <- "https://www.cell.com/cms/10.1016/j.cels.2016.09.002/attachment/ddfaa172-dc06-4228-b93e-b385300c5cfe/mmc3.xlsx"
table_s3 <- file.path(docs_dir, "extdata", "mmc3.xlsx")

get_file(url, table_s3)
sheets <- readxl::excel_sheets(table_s3)
sup_dat <- map(sheets, ~readxl::read_excel(table_s3, sheet = .x))
names(sup_dat) <- sheets


sup_dat_sig_genes <- map(sup_dat, 
                         ~filter(.x, padj == 0) %>% 
  arrange(log2FoldChange) %>% 
  select(X__1, log2FoldChange) %>% 
  mutate(log2FoldChange = 2^(-log2FoldChange)) %>% 
  dplyr::slice(1:100))

sup_dat_all_sig_genes <- unique(bind_rows(sup_dat_sig_genes)$X__1)
```


```{r}

shared_genes <- intersect(sup_dat_all_sig_genes,
                          rownames(bcells@data))

sc_signatures <- map2(sup_dat_sig_genes, 
     names(sup_dat_sig_genes), 
     function(x, y){ 
       df <- filter(x, X__1 %in% shared_genes) %>% 
         as.data.frame() %>% 
         tibble::column_to_rownames("X__1")
       colnames(df) <- y
       as.matrix(df)})

b <- ifelse(as.matrix(bcells@data[shared_genes, ]) > 0,
            0,
            1)
b <- bcells@data[shared_genes, ]
res <- map(sup_dat_sig_genes,
           ~compare_lists(b, .x, n = 10000, metric = "jaccard")) 

res <- map(sup_dat_sig_genes,
           function(x){
             genes <- intersect(rownames(x),
                                shared_genes)
             run_cor(b, 
                    bcells@meta.data, 
                    x[genes, ], 
                    genes,
                    cluster_col = "res.0.3")}) 

res <- do.call(cbind, res)

bcell_meta <- GetCellEmbeddings(bcells, "tsne") %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell")

plts <- plot_cor(res,
         bcell_meta,
         colnames(res), cluster = "cell")

plt <- plot_grid(plotlist = plts, nrow = 4)
save_plot("cor_based_on_scrna_list_.pdf", plt,
          nrow = 3,
          ncol = 4)
```


```{r}


url <- "https://www.cell.com/cms/10.1016/j.cmet.2016.08.020/attachment/6a140600-6b84-46f7-abc4-65058fb9e587/mmc3.xlsx"
table_s2 <- file.path(docs_dir, "extdata", "mmc3_sandberg.xlsx")

get_file(url, table_s2)
sheets <- readxl::excel_sheets(table_s2)
sup_dat <- map(sheets, ~readxl::read_excel(table_s2, sheet = .x))
names(sup_dat) <- sheets

dat <- sup_dat$`cell type specific`
dat <- dat[, c(2, 4, 6, 8, 10, 12)]
colnames(dat) <- dat[2, ]
dat <- dat[-1, ]
dat <- dat[-1, ]
dat <- dat[-1, ]
dat <- dat %>% 
  gather(cluster, gene) %>% 
  na.omit() %>% 
  mutate(gene = str_replace_all(gene, "'", "")) 

out <- map(split(dat, dat$cluster),
    function(x){
      df <- data_frame(x$gene)
      colnames(df) <- unique(x$cluster)
      df})

out_all_genes <- unique(unname(unlist(out)))
shared_genes <- intersect(out_all_genes, rownames(bcells@data))
out <- map(out, ~.x[.x[, 1, drop = T] %in% shared_genes, ])

b <- ifelse(as.matrix(bcells@data[shared_genes, ]) > 0,
            0,
            1)

res <- map(out,
           ~compare_lists(b, .x, n = 10000)) 
res <- do.call(cbind, res)

bcell_meta <- GetCellEmbeddings(bcells, "tsne") %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell")

colnames(res) <- c("acinar", 
                   "ductal",
                   "alpha",
                   "beta",
                   "gamma",
                   "delta")
plts <- plot_cor(res,
         bcell_meta,
         colnames(res), cluster = "cell")

plt <- plot_grid(plotlist = plts, nrow = 4)
save_plot("cor_based_on_scrna_list2.pdf", plt,
          nrow = 3,
          ncol = 4)
```